### ISSUE TITLE
Does codespell support skipping the presence of a word in a specific file, but still search for it elsewhere?

### ISSUE THREAD

**Issue by ScriptAutomate** [codespell-project/codespell#1601]
Documentation for codespell seems to say that I can skip entire files or directories, or words, but not a word found in a specific file.

Example:

If documenting how codespell works, and showing the output involving an incorrectly spelled word, you would like to skip that misspelled word being used in the example of the file.

Otherwise, I'd have to skip the entire file.

If this is a supported feature already, I'd love to know what needs to be done. I imagine it could be an extension of `codespell -L`, perhaps another argument that becomes available when using `-L`. Unless there is another suggestion?

**Comment by peternewman** [comment]
From the help, this is your best bet:
```
  -x FILE, --exclude-file FILE
                        FILE with lines that should not be checked for errors
                        or changed
```

**Comment by ScriptAutomate** [comment]
> From the help, this is your best bet:
> 
> ```
>   -x FILE, --exclude-file FILE
>                         FILE with lines that should not be checked for errors
>                         or changed
> ```

Yeah, I was thinking the same. Some kind of combination of `-x` and `-L`. Though, when using `-x` alone, this excludes an entire file from codespell, much like `-S`, and I can't tell the difference between `-x` and `-S`. One is skip a file (or files) while the other is exclude a file. Is it just that `-S` accepts more than one name, and globbing?

It looks like I can achieve what I am looking for by doing the following (as I am using CI):

```
codespell -x example.md  # Run against all files except example.md
codespell -L <word(s)-to-skip> example.md  # Run against example.md, but skip single intentional misspelling
```

I'm using `pre-commit`, and wanted to see if there was a single command that achieves both of those: run against all files, but with `example.md`, ignore a particular misspelling (but check for that misspelling against all other files still).

**Comment by peternewman** [comment]
Though, when using `-x` alone, this excludes an entire file from codespell, much like `-S`, and I can't tell the difference between `-x` and `-S`.

You've misunderstood, `-x` is a file with a list of lines to ignore across all files, whereas -S as you say is to skip.

So if one of your files looked like this:
```
This is a test file
This line has a deliberate speling mistake
For testing codespell
```

The file you pass into -x (or part of it), would look like this:
```
This line has a deliberate speling mistake
```

So it only ignores that one instance of the word `speling`. Other ones in other files would get flagged.

And a real world example:
https://github.com/OpenLightingProject/rdm-app/blob/master/.codespellignore

So any place where that specific line (or other lines in your `-x` file) are found, the whole line gets ignored, if you had for example also missed that you'd typed `Ths` it wouldn't warn you, or if we later add and spot another typo you wouldn't know.

As this is clearly a bit confusing, we'd welcome any suggestions for how to better reword that bit of help. Perhaps "FILE with a list of lines" or "multiple" or "various"? Or "... within any files being checked"?

**Comment by dg0yt** [comment]
I can't help with the wording. But I would rather prefer to specify ignored phrases instead of full lines. I'm runnning into `Water trough` getting flagged `trough ==> through`. And I have multiple occurences of `Childs[...]` flagged `Childs ==> Children, child's`

**Comment by ScriptAutomate** [comment]
For others who find this issue, this worked for me in my `.pre-commit-config.yaml` file:

```yaml
  -   repo: https://github.com/codespell-project/codespell
      rev: v2.0.0
      hooks:
      - id: codespell
        description: Base spellcheck across repository.
        args: ['--exclude-file', '.codespellignore']
```

`-x` is the same as `--exclude-file`, and seemed like a more readable approach to my config for others viewing.

As mentioned by @peternewman, it meant using `.codespellignore` as a file to store _exact lines of text to skip when using `codespell`_. There currently is not a way to skip the flagging of a specific _word_, but only of a sentence (which would be containing the text you wish `codespell` to ignore). As he also mentions, this does mean that **all misspellings** in the line of text will be skipped by `codespell`, so care should be taken when doing this.

This may cause confusion in a repo with multiple contributors if a sentence is edited but still contains the word that `codespell` borks out on.

I might put in a PR with the above information, since I seem to now understand these args a bit more now that I have taken a look.

**Comment by matkoniecz** [comment]
`--exclude-file` excludes entire file, `--ignore-words-list` ignores word everywhere so neither work for "exclude specific word only in a specific file".

In my case where affected files are stable I ended using

```
codespell --bunch-of-codespell-arguments
|grep -v "StreetComplete/app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/PolylinesSerializer.kt:41: ois ==> is" \
|grep -v "StreetComplete/app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/PolylinesSerializer.kt:42: ois ==> is" \
|grep -v "StreetComplete/app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/PolylinesSerializer.kt:43: ois ==> is" \
|grep -v "StreetComplete/app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/PolylinesSerializer.kt:44: ois ==> is" \
|cat
```

**Comment by peternewman** [comment]
> I can't help with the wording. But I would rather prefer to specify ignored phrases instead of full lines. I'm runnning into `Water trough` getting flagged `trough ==> through`.

I assume you don't want to just ignore all trough's @dg0yt in which case you could skip that word or the whole rare dictionary? You could probably also abuse the regex URL filter to make that phrase vanish and hence not show up.

> And I have multiple occurences of `Childs[...]` flagged `Childs ==> Children, child's`

Sorry I don't follow this, `childs foo` isn't valid, or do you have a variable called childs?

So it strikes me now we've got the config files, perhaps the easiest thing people could do is make use of them; have your main config stored in them, and run codespell twice, once skipping your problematic files, and once just checking those files but excluding the problematic word(s).

> `--exclude-file` excludes entire file

No it doesn't, that's what skip does. Please see the man file/help. This just excludes lines matching the contents of lines in the exclude file.

> In my case where affected files are stable I ended using
> 
This would be far simpler and zero maintenance (unless someone types ois in a comment in that file), but for that case use --exclude-file instead:
`grep -v "StreetComplete/app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/PolylinesSerializer.kt:[0-9]+: ois ==> is"`

> Main negative effect is losing colors in display

Grep should pass them on, you might need `--color=always`

**Comment by matkoniecz** [comment]
> No it doesn't, that's what skip does. Please see the man file/help

sorry, I was confused! I actually have seen it but managed to become confused.

Though `--exclude-file` is a bit confusingly named given "exclude file" vs "file with line specific exclusions". Though I guess that changing name now wold not be a good idea.

> So it strikes me now we've got the config files, perhaps the easiest thing people could do is make use of them; have your main config stored in them, and run codespell twice, once skipping your problematic files, and once just checking those files but excluding the problematic word(s).

That requires maintaining per project list of files and looking them up to run codespell. And always checking before running codespell is there some project-specific exclusion list.

Other option that I see is to have global codespell alias that will always 
- scan current directory
- scan all files, across entire system which with some excluded words. Multiple commands required, one for each combination of skipped words

**Comment by peternewman** [comment]
> Though `--exclude-file` is a bit confusingly named given "exclude file" vs "file with line specific exclusions". Though I guess that changing name now wold not be a good idea.

If we can improve the help wording further then great, as you say otherwise migrating to another command line argument is probably more hassle than it's worth.

> That requires maintaining per project list of files and looking them up to run codespell. And always checking before running codespell is there some project-specific exclusion list.

Which is pretty trivial via CI or precommit.

> Other option that I see is to have global codespell alias that will always
> 
>     * scan current directory
> 
>     * scan all files, across entire system which with some excluded words. Multiple commands required, one for each combination of skipped words

I suspect a far easier option would be some sort of typo to exclude file mapping system, e.g.:
```
ois=>StreetComplete/app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/PolylinesSerializer.kt, foo.txt
crate=>bar.rs
```

**Comment by matkoniecz** [comment]
> Which is pretty trivial via CI or precommit.

I have a bit different use case then - I scan entire folder with all source code, rather than setting CI/precommit on each of 200+ repositories, including ones not controlled by me.

And plan to extend it to all plaintext files (if I figure out way to avoid scanning Polish text).

**Comment by peternewman** [comment]
> > Which is pretty trivial via CI or precommit.
> 
> I have a bit different use case then - I scan entire folder with all source code, rather than setting CI/precommit on each of 200+ repositories, including ones not controlled by me.

Yeah that's different! What do you do with the results, fix and push back? I've always primarily run it (and spellintian) as CI tools.

> And plan to extend it to all plaintext files (if I figure out way to avoid scanning Polish text).

We could split the dictionaries further, or you could have your own custom dictionary which removes any typos which are valid Polish words. But that would be another good use case for having config per project and so only ignoring Polish words in Polish projects.

> > I suspect a far easier option would be some sort of typo to exclude file mapping system, e.g.:
> 
> that looks nice!

Although you've got the usual how do you do your escaping etc, I guess having the RHS being a glob might be easiest/more flexible.

**Comment by matkoniecz** [comment]
> Yeah that's different! What do you do with the results, fix and push back? I've always primarily run it (and spellintian) as CI tools.

Exactly. See say https://github.com/florianfesti/boxes/commit/62b30e3d88157de5ecaa93fea1e88396392aa903 (this projects has many more typos) or this StreetComplete commit. But most are in my own small projects or notes.

> We could split the dictionaries further, or you could have your own custom dictionary which removes any typos which are valid Polish words. But that would be another good use case for having config per project and so only ignoring Polish words in Polish projects.

I vaguely imagined detecting language automatically, per file. That would help with translation files where assigning them manually in each project would be irritating. But building typo list is a significant effort anyway.

**Comment by peternewman** [comment]
> Exactly. See say [florianfesti/boxes@62b30e3](https://github.com/florianfesti/boxes/commit/62b30e3d88157de5ecaa93fea1e88396392aa903) (this projects has many more typos) or this StreetComplete commit. But most are in my own small projects or notes.

I see. Personally I'd open a PR to run codespell as an action against this repo, as otherwise new typos will just appear in future; not much point just cleaning it once arbitrarily.

> I vaguely imagined detecting language automatically, per file. That would help with translation files where assigning them manually in each project would be irritating. But building typo list is a significant effort anyway.

I don't know how you'd do that, I guess you could count matching words in the document to corrections in each dictionary to try and identify which dictionary to use, but I don't know how effective that would be.

**Comment by dg0yt** [comment]
> > I can't help with the wording. But I would rather prefer to specify ignored phrases instead of full lines. I'm runnning into `Water trough` getting flagged `trough ==> through`.
> 
> I assume you don't want to just ignore all trough's @dg0yt in which case you could skip that word or the whole rare dictionary? You could probably also abuse the regex URL filter to make that phrase vanish and hence not show up.

I'm looking for a *clean* solution to ignore certain phrases. Something between single words and full lines.

> > And I have multiple occurences of `Childs[...]` flagged `Childs ==> Children, child's`
> 
> Sorry I don't follow this, `childs foo` isn't valid, or do you have a variable called childs?

I do use another library that uses `Childs` in its API.

**Comment by peternewman** [comment]
> > > I can't help with the wording. But I would rather prefer to specify ignored phrases instead of full lines. I'm runnning into `Water trough` getting flagged `trough ==> through`.
> > 
> > 
> > I assume you don't want to just ignore all trough's @dg0yt in which case you could skip that word or the whole rare dictionary? You could probably also abuse the regex URL filter to make that phrase vanish and hence not show up.
> 
> I'm looking for a _clean_ solution to ignore certain phrases. Something between single words and full lines.

I forgot, we've got a plain regex filter too, added in #1592 . I told you it would come in useful @jonmeow !

Make it match your phrase and it will be like it doesn't exist so it won't be reported for typos!

> > > And I have multiple occurences of `Childs[...]` flagged `Childs ==> Children, child's`
> > 
> > 
> > Sorry I don't follow this, `childs foo` isn't valid, or do you have a variable called childs?
> 
> I do use another library that uses `Childs` in its API.

Ah gotcha, that makes sense, if unfortunate that it's mispelt!

**Comment by akimd** [comment]
It is very important to me to be able to ignore a specific word in a specific file, and nowhere else.  So I use the following wrapper to make a glorified version of grep-based ignore list above.  And I'm a Ruby programmer, so you'll have to use your favorite ruby2python.

```ruby
#! /usr/bin/env ruby
# frozen_string_literal: true

# codespell is nice, but does not allow to specify specific
# exceptions: you can ignore words, or files, but not specific words
# in specific files.
#
# This wrapper implements this.  See .codespell.ignore, whose format
# is self-explanatory:
#
# ```
# doc/misc.md: ist
# # In JSON, "\nIn" or "\nWe".
# tests/**/*.json: nIn|nWe
# ```

require 'colorize'
require 'open3'
require 'pathname'

# [[GLOB, REGEX], ...]
ignores =
  Pathname('.codespell.ignores')
    .read.lines(chomp: true)
    .reject { |it| it.empty? || it.start_with?('#') }
    .map { |it| files, re = it.split(/:\s*/, 2); [files, /#{re}/] }

out, _err, _status = Open3.capture3('codespell', *ARGV)

errors = out.lines(chomp: true).filter_map { |line|
  # doc/misc.md:183: ist ==> is, it, its, it's, sit, list
  if (file, line, word, rest = line.match(%r{([\w\-./]+):(\d+): (.*?) (==> .*)})&.captures)
    accepted = ignores.any? { |files, words|
      File.fnmatch?(files, file, File::FNM_EXTGLOB) && words.match?(word)
    }
    "#{file.yellow}:#{line.yellow}: #{word.red} #{rest}" if !accepted
  end
}

if errors.any?
  puts errors
  exit 1
end
